### PR TITLE
Feature/plugin optimization

### PR DESCRIPTION
**Pull request by silvexis**
Address bugs in plugins and VPC flow logs and improve overall stability of the system under load

### CODE REVIEW

File: reactor/database/models/dynamodb/cz_reactor_environments.py [hunk 1]
```diff
@@ -64,7 +81,7 @@ def create_id():
 
     def _clear_caches(self):
         try:
-            self.get_from_account_id_cached.cache_clear()
+            # self.get_from_account_id_cached.cache_clear()
```
Comment (tiyberius) [inline reactor/database/models/dynamodb/cz_reactor_environments.py]: remove this?

File: reactor/database/models/dynamodb/cz_reactor_environments.py [hunk 2]
```diff
@@ -78,18 +95,20 @@ def get_all(self):
...
-    @functools.lru_cache(maxsize=512)
-    def get_from_account_id_cached(self, account_id):
-        """Get environment for given account_id. This method is memoized. Client needs
-        to be aware if it is getting/putting/getting often and should use the `cache_info`
-        and `cache_clear` methods provided by `functools.lru_cache`.
+    def get_from_account_id(self, account_id):
+        """
+        Get environment for given account_id.
+
+        A lru cached version of this method is also available as the function `get_env_from_account_id_cached`
```
Comment (tiyberius) [inline reactor/database/models/dynamodb/cz_reactor_environments.py]: okay, this is weird. we shouldn't have two copies of this code in the codebase

File: reactor/database/models/dynamodb/cz_reactor_environments.py [hunk 3]
```diff
@@ -136,7 +153,7 @@ def remove(self, account_id):
         Ret:
             success (bool): success of delete operation
         """
-        self._clear_caches()
+        # self._clear_caches()
```
Comment (tiyberius) [inline reactor/database/models/dynamodb/cz_reactor_environments.py]: remove?

File: reactor/features/ingest.py [hunk 4]
```diff
@@ -65,7 +66,9 @@ def ingest_events(events, namespace, client_details=None):
     for event in events:
         if event['EventSource'] == 'aws:sns':
             connected_account_id = event['EventSubscriptionArn'].split(":")[4]
-            environment = environments.get_from_account_id_cached(connected_account_id)
+            # environment = environments.get_from_account_id_cached(connected_account_id)
```
Comment (tiyberius) [inline reactor/features/ingest.py]: remove?

File: reactor/features/ingest.py [hunk 5]
```diff
@@ -166,7 +169,10 @@ def process_aws_cloudtrail_event(queue_name, reactor_account_id, connected_accou
         s3 = Provider(service='s3', role_arn=role_arn, external_id=external_id, endpoint_url=s3_endpoint_url)
 
         logger.debug("S3 Role: {} {}".format(role_arn, external_id))
-        environment = environments.get_from_account_id_cached(connected_account_id)
+        # environment = environments.get_from_account_id_cached(connected_account_id)
```
Comment (tiyberius) [inline reactor/features/ingest.py]: remove?

File: reactor/features/resources/curate.py [hunk 6]
```diff
@@ -98,6 +127,8 @@ def process_new_resource(resource_id, timestamp, service_type, environment_id, d
         resource_id (str): ID of the tracking record to process
         service_type (str): AWS standard service type attached to the resource
         environment_id (str): CloudZero Environment ID
+        cz_metadata (dict):
+        aws_account_details (dict):
```
Comment (tiyberius) [inline reactor/features/resources/curate.py]: can we fill this out

File: reactor/features/resources/plugins/ec2.py [hunk 7]
```diff
@@ -27,29 +28,59 @@ def enrich_resource(event):
         event: dict containing details of the event from AWS in the format
                {'resource_id': <RESOURCE ID>,
                 'timestamp': <TIMESTAMP>,
-                'environment_id': <ENV ID>}
+                'environment_id': <ENV ID>,
+                'aws_account_details': dict,        # AWS connection information
```
Comment (tiyberius) [inline reactor/features/resources/plugins/ec2.py]: why does this need a comment explaining what it is ... should the key be something different?

File: reactor/features/resources/plugins/base.py [hunk 8]
```diff
@@ -98,21 +99,7 @@ def __init__(self, resource_id, timestamp, environment_id) -> None:
...
-
-        try:
-            self.aws = Provider(role_arn=self.aws_account_details['role_arn'],
-                                external_id=self.aws_account_details['external_id'],
-                                region=self.region or 'us-east-1',
-                                service=self.service_type)
-        except (botocore.exceptions.UnknownServiceError, OSError):
-            logger.warning(f'Resource | Mostly harmless attempt to create '
-                           f'AWS service connection to {self.service_type}')
-            self.aws = None
```
Comment (tiyberius) [inline reactor/features/resources/plugins/base.py]: cool, so sometimes we don't even need this connection

File: reactor/features/resources/plugins/ec2.py [hunk 9]
```diff
@@ -27,29 +28,59 @@ def enrich_resource(event):
...
-            data, resource_state, resource_status = process_vpc(resource)
+            data, resource_state, resource_status = process_vpc(resource, ec2_provider)
+            flow_log_status = cz_metadata.get('flow_log_status')
+            if not flow_log_status:
+                logger.info(f"Attempting to configure VPC Flow logs for {resource.id}")
+                flow_log_group_name, subscribe_success = configure_flow_log(ec2_provider, resource.id)
+                resource.update_cz_metadata('flow_log_status', {"log_group_name": flow_log_group_name,
+                                                                "subscribed": subscribe_success,
+                                                                "timestamp": datetime.now(
+                                                                    tz=timezone.utc).isoformat()})
```
Comment (tiyberius) [inline reactor/features/resources/plugins/ec2.py]: not sure I like the placement of this, I think we should try to keep this `if` `elif` as clean as possible, but won't hold up the PR

File: reactor/features/resources/plugins/ec2.py [hunk 10]
```diff
@@ -69,27 +100,28 @@ def enrich_resource(event):
...
 
-def process_instance(resource):
+def process_instance(resource, aws):
     data = {}
+    instance = None
     try:
-        instance = list(resource.aws.resource.instances.filter(InstanceIds=[resource.id]))[0]
+        instance = list(aws.resource.instances.filter(InstanceIds=[resource.id]))[0]
+        # instance = aws.resource.Instance(resource.id)
+        # instance.load()
```
Comment (tiyberius) [inline reactor/features/resources/plugins/ec2.py]: remove?

File: test/integration/features/resources/plugins/test_base.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+                                                tz=timezone.utc).isoformat()})
+
+    updated_resource = sample_resource_database.table.query(
+        KeyConditionExpression=Key('res_id').eq(all_resources[0]['res_id']) & Key('timestamp').eq(
+            all_resources[0]['timestamp'])
+    )
+
+    assert updated_resource['Items'][0]
+    assert updated_resource['Items'][0]['cz_metadata']['nothing']
+    assert updated_resource['Items'][0]['cz_metadata']['something']['exciting_data'] == "test_data"
```
Comment (tiyberius) [inline test/integration/features/resources/plugins/test_base.py]: what is all this code

File: reactor/aws/cloudtrail.py [hunk 12]
```diff
@@ -258,7 +261,8 @@ def normalize(self, environments, event_owner_id=None):
         event_datetime = ciso8601.parse_datetime(raw_event['eventTime'])
         event_timestamp = Decimal(event_datetime.timestamp())
 
-        environment = environments.get_from_account_id_cached(raw_event.get('recipientAccountId'))
+        environment = get_env_from_account_id_cached(raw_event.get('recipientAccountId'))
+        environment = environments.get_from_account_id(raw_event.get('recipientAccountId'))
```
[the remark this answers is not in the document]
Comment (imcgaunn) [reply reactor/aws/cloudtrail.py]: I echo tobias' question.

File: reactor/database/models/dynamodb/cz_reactor_resources.py [hunk 13]
```diff
@@ -134,7 +134,19 @@ def query_by_resource_name(self, object_name):
...
             if 'ResourceNotFoundException' in str(err):
-                raise common.DataNotFound('Environment id does not exist')
+                raise common.DataNotFound('Resource does not exist')
+            else:
+                logger.error('ClientError while trying to get environment_id {}: {}'.format(object_name, err))
+                raise
+
+    def query_by_resource_name_with_filter(self, object_name, filter_expression):
+        try:
+            return self.table.query(IndexName='ResourceNameGSI',  # THIS ID IS SET IN SERVERLESS.YML
```
Comment (imcgaunn) [inline reactor/database/models/dynamodb/cz_reactor_resources.py]: This should be a named constant somewhere.

File: reactor/features/ingest.py [hunk 14]
```diff
@@ -247,15 +254,18 @@ def process_aws_cloudtrail_event(queue_name, reactor_account_id, connected_accou
...
-            events, unique_resource_ids = normalize_cloudtrail_events(data_records, environments, event_stats,
-                                                                      message_source_environment_id)
-            write_events_to_bucket(events, message_source_environment_id, namespace, reactor_account_id, s3_provider)
             resources.create_new_resources(unique_resource_ids, message_source_environment_id, connected_account_id)
             message.delete()  # Message is processed, safe to remove from SQS
 
         except Exception as err:
+            logger.error(f'Failed to create {len(unique_resource_ids)} resources, skipping resource create')
             logger.exception("{} : {}".format(type(err), err))
+            message.delete()  # we delete the message because we don't want to get stuck in a loop
```
Comment (imcgaunn) [inline reactor/features/ingest.py]: I would rephrase this comment (or remove it).

File: reactor/features/resources/curate.py [hunk 15]
```diff
@@ -12,22 +12,41 @@
...
+    environments = dynamodb.CZReactorEnvironments()
+    environment = environments.get_from_environment_id(environment_id)
+    aws_account_details = environment['account_details']
+    return aws_account_details
+
+
+def get_cz_metadata(resource_name, dbh):
+    filter_expression = Attr('disposition').ne('pending')
+    metadata = dbh.query_by_resource_name_with_filter(resource_name, filter_expression)
+    if metadata and len(metadata) == 1:
```
Comment (imcgaunn) [inline reactor/features/resources/curate.py]: this looks like a good candidate for refactoring using Exceptions for control flow.

why not just:

try:
  return metadata[0].get('cz_metadata', {})
except TypeError: (not sure if that's the right exception)
  return {}

Comment (imcgaunn) [reply reactor/features/resources/curate.py]: I mean replace the `if / else` with the code above.

File: reactor/features/resources/curate.py [hunk 16]
```diff
@@ -149,7 +180,7 @@ def write_resource_to_s3(resource_id, environment_id, resource):
     resource_object_value = utils.encode_and_compress(json.dumps(resource))
     logger.debug(f'About to write to the "{bucket_name}" bucket under the key "{resource_object_key}": {resource}')
     result = Provider(service='s3').client.put_object(Bucket=bucket_name, Key=resource_object_key,
-                                                      Body=resource_object_value)
+                                                      Body=resource_object_value, ServerSideEncryption='AES256')
```
Comment (imcgaunn) [inline reactor/features/resources/curate.py]: does this resolve the TODO from above, or is the TODO above because this part doesn't work?

File: reactor/features/resources/plugins/ec2.py (hunk 7 above)
Comment (imcgaunn) [reply reactor/features/resources/plugins/ec2.py]: seconded. It seems like `cz_metadata` is specific to resources. Why not call it `cz_resource_metadata` or just `resource_metadata`?

File: reactor/features/resources/plugins/ec2.py [hunk 17]
```diff
@@ -69,27 +99,28 @@ def enrich_resource(event):
...
-        create_vpc_resources_from_ec2_instance(instance, resource)
-
-        create_eni_resources_from_ec2_instance(instance, resource)
-
     except ClientError as error:
-        logger.debug(error)
         if 'InvalidInstanceID.NotFound' in str(error):
-            logger.info(f"Instance {resource.resource_id} went away")
-            resource_status = reactor.constants.RESOURCE_STATUS_REMOVED
+            logger.debug(f"Instance {resource.resource_id} went away {error}")
```
Comment (imcgaunn) [inline reactor/features/resources/plugins/ec2.py]: is there an issue with this remaining at `info` level? This seems like a useful informational message.

File: reactor/features/resources/plugins/ec2.py [hunk 18]
```diff
@@ -150,97 +181,91 @@ def process_vpc(resource):
...
-            # To make the process of event driven CloudWatch Logs delivery possible we will need to implement a
-            # "Logical Destination" which is a CloudWatch Logs system for allowing cross account delivery of events
-            # to a Kinesis stream.
-            # https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutSubscriptionFilter.html
-
-            get_flow_log_for_vpc_cached.cache_clear()
+        logs_provider = Provider(service='logs', **ec2_provider.aws_connection_details)
+        flow_log_group_name = create_vpc_flow_log(vpc_id, logs_provider, ec2_provider)
+        subscribe_success = subscribe_to_flow_log(vpc_id, logs_provider)
+        # This attempts to subscribe our VPC Flow Log to a reactor ingest function but unfortunately
```
Comment (imcgaunn) [inline reactor/features/resources/plugins/ec2.py]: this is a nice comment to have! 👍

File: serverless.yml [hunk 19]
```diff
@@ -119,7 +116,7 @@ functions:
     memorySize: 2048
   resources-enrich-plugin-ec2:
     handler: lambda_handlers.resources__enrich_plugin_ec2
-    memorySize: 2048
+    memorySize: 512
```
Comment (imcgaunn) [inline serverless.yml]: what's the reason for this decrease? Were we not using the extra resources?

File: reactor/aws/cloudtrail.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (silvexis) [reply reactor/aws/cloudtrail.py]: oops

File: reactor/database/models/dynamodb/cz_reactor_environments.py (hunk 2 above)
Comment (silvexis) [reply reactor/database/models/dynamodb/cz_reactor_environments.py]: fixed, the cached will call the uncached ver.

File: reactor/features/resources/plugins/base.py (hunk 8 above)
Comment (silvexis) [reply reactor/features/resources/plugins/base.py]: sometimes...

File: reactor/features/resources/plugins/ec2.py (hunk 7 above)
Comment (silvexis) [reply reactor/features/resources/plugins/ec2.py]: i'll remove the comment ;)

File: test/integration/features/resources/plugins/test_base.py (hunk 11 above)
Comment (silvexis) [reply test/integration/features/resources/plugins/test_base.py]: i don't even know who I am anymore

File: serverless.yml (hunk 19 above)
Comment (silvexis) [reply serverless.yml]: correct, I tested with a number of different sizes as well and this seemed to offer reasonable performance. For reference, the actual memory usage is around 100mb but setting this number to less than 512 resulting in a 5x increase in execution time

File: reactor/database/models/dynamodb/cz_reactor_environments.py [hunk 20]
```diff
@@ -19,18 +19,7 @@
...
-            return result
-        else:
-            raise common.DataNotFound('Account id does not exist')
-    except botoerr.ClientError as err:
-        if 'ResourceNotFoundException' in str(err):
-            raise common.DataNotFound('Account id does not exist')
-        else:
-            logger.error('ClientError while trying to get account_id {}: {}'.format(account_id, err))
-            raise
+    return environments.get_from_account_id(account_id)
```
Comment (tiyberius) [inline reactor/database/models/dynamodb/cz_reactor_environments.py]: 👍
